File: modules/device-diagnostics/ios/DeviceDiagnosticsModule.swift
import ExpoModulesCore
import UIKit

struct VibrationOptions: Record {
  @Field var duration: Int = 500
  @Field var intensity: Double = 1.0
}

public class DeviceDiagnosticsModule: Module {
  private var batteryObservers: [NSObjectProtocol] = []
  private var isManuallyMonitoring = false

  public func definition() -> ModuleDefinition {
    Name("DeviceDiagnostics")

    Events("batteryStateChanged")

    OnStartObserving("batteryStateChanged") { [weak self] in
      self?.startBatteryMonitoring()
    }

    OnStopObserving("batteryStateChanged") { [weak self] in
      self?.stopBatteryMonitoring()
    }

    OnDestroy { [weak self] in
      self?.stopBatteryMonitoring()
    }

    Function("getPlatformName") {
      "ios"
    }

    Function("getDeviceInfo") {
      let device = UIDevice.current
      return [
        "platform": "ios",
        "model": device.model,
        "systemVersion": device.systemVersion
      ]
    }

    Function("vibrate") { (options: VibrationOptions) throws in
      guard options.duration > 0 else {
        throw Exception(name: "InvalidArgumentException", description: "`duration` must be greater than 0.")
      }
      guard options.intensity >= 0 && options.intensity <= 1 else {
        throw Exception(name: "InvalidArgumentException", description: "`intensity` must be between 0 and 1.")
      }

      let style: UIImpactFeedbackGenerator.FeedbackStyle
      switch options.duration {
      case ..<200:
        style = .light
      case 200..<500:
        style = .medium
      default:
        style = .heavy
      }

      DispatchQueue.main.async {
        let generator = UIImpactFeedbackGenerator(style: style)
        generator.prepare()
        if #available(iOS 13.0, *) {
          generator.impactOccurred(intensity: CGFloat(options.intensity))
        } else {
          generator.impactOccurred()
        }
      }
    }

    AsyncFunction("getAvailableDiskSpace") { () throws -> [String: Int64] in
      let attributes = try FileManager.default.attributesOfFileSystem(forPath: NSHomeDirectory())
      guard let freeSize = attributes[.systemFreeSize] as? NSNumber else {
        throw Exception(name: "DiskSpaceUnavailableException", description: "Available disk space is unavailable.")
      }

      return ["availableBytes": freeSize.int64Value]
    }

    Function("startMonitoring") {
      self.isManuallyMonitoring = true
      self.startBatteryMonitoring()
    }

    Function("stopMonitoring") {
      self.isManuallyMonitoring = false
      self.stopBatteryMonitoring()
    }
  }

  private func startBatteryMonitoring() {
    UIDevice.current.isBatteryMonitoringEnabled = true
    guard batteryObservers.isEmpty else {
      emitBatteryStateIfAvailable()
      return
    }

    let center = NotificationCenter.default
    let levelObserver = center.addObserver(
      forName: UIDevice.batteryLevelDidChangeNotification,
      object: nil,
      queue: .main
    ) { [weak self] _ in
      self?.emitBatteryStateIfAvailable()
    }
    let stateObserver = center.addObserver(
      forName: UIDevice.batteryStateDidChangeNotification,
      object: nil,
      queue: .main
    ) { [weak self] _ in
      self?.emitBatteryStateIfAvailable()
    }

    batteryObservers = [levelObserver, stateObserver]
    emitBatteryStateIfAvailable()
  }

  private func stopBatteryMonitoring() {
    let center = NotificationCenter.default
    batteryObservers.forEach { center.removeObserver($0) }
    batteryObservers.removeAll()
    UIDevice.current.isBatteryMonitoringEnabled = false
  }

  private func emitBatteryStateIfAvailable() {
    let device = UIDevice.current
    let level = device.batteryLevel
    guard level >= 0 else {
      return
    }

    let isCharging = device.batteryState == .charging || device.batteryState == .full
    sendEvent("batteryStateChanged", [
      "level": level,
      "isCharging": isCharging
    ])
  }
}

File: modules/device-diagnostics/android/src/main/java/expo/modules/devicediagnostics/DeviceDiagnosticsModule.kt
package expo.modules.devicediagnostics

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class VibrationOptions : Record {
    @Field
    var duration: Long = 500

    @Field
    var intensity: Double = 1.0
}

class DeviceDiagnosticsModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("DeviceDiagnostics")

        Function("getPlatformName") {
            "android"
        }

        Function("getDeviceInfo") {
            mapOf(
                "platform" to "android",
                "model" to Build.MODEL,
                "systemVersion" to Build.VERSION.RELEASE
            )
        }

        Function("vibrate") { options: VibrationOptions ->
            if (options.duration <= 0) {
                throw IllegalArgumentException("`duration` must be greater than 0.")
            }
            if (options.intensity < 0.0 || options.intensity > 1.0) {
                throw IllegalArgumentException("`intensity` must be between 0 and 1.")
            }

            val context = appContext.reactContext
                ?: throw IllegalStateException("React context is unavailable.")

            val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
                context.getSystemService(VibratorManager::class.java)?.defaultVibrator
            } else {
                @Suppress("DEPRECATION")
                context.getSystemService(Context.VIBRATOR_SERVICE) as? Vibrator
            }

            if (vibrator == null || !vibrator.hasVibrator()) {
                throw IllegalStateException("Vibrator service is unavailable on this device.")
            }

            val amplitude = (options.intensity * 255).toInt().coerceIn(1, 255)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                vibrator.vibrate(VibrationEffect.createOneShot(options.duration, amplitude))
            } else {
                @Suppress("DEPRECATION")
                vibrator.vibrate(options.duration)
            }
        }
    }
}
